### PR TITLE
District2

### PR DESCRIPTION
**Pull request by Yoel-Perman-Brilliant**

### CODE REVIEW

File: src/main/java/frc/robot/OI.java [hunk 1]
```diff
@@ -39,8 +39,31 @@ public OI() {
...
+                return false;
+            }
+        });
+        xbox.getYellowButton().whileHeld(new MoveGenericSubsystem(climberWinch, ClimberWinch.UP_SPEED) {
+            @Override
+            public boolean isFinished() {
+                return false;
+            }
+        });
         xbox.getBlueButton().whenPressed(new MoveGenericSubsystem(climberWinch, 0));
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/OI.java:67]: override `isFinished` to return `true`

File: src/main/java/frc/robot/OI.java [hunk 2]
```diff
@@ -39,8 +39,31 @@ public OI() {
...
         xbox.getRedButton().whenPressed(new MoveGenericSubsystem(roller, 0));
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/OI.java:68]: ditto

File: src/main/java/frc/robot/commands/IntakeCargo.java [hunk 3]
```diff
@@ -11,12 +11,10 @@
 
     public IntakeCargo() {
         IntakeRoller intakeRoller = IntakeRoller.getInstance();
-        IntakePlacer intakePlacer = IntakePlacer.getInstance();
         Transfer transfer = Transfer.getInstance();
         IntakeToTransfer intakeToTransfer = IntakeToTransfer.getInstance();
-        addRequirements(intakeRoller, intakePlacer, transfer, intakeToTransfer);
         addCommands(
-                new MoveIntakePlacerDown(),
+                new MoveIntakePlacerDown().withTimeout(4),
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/IntakeCargo.java:17]: https://www.youtube.com/watch?v=aJnNltIVzc8

File: src/main/java/frc/robot/Robot.java [hunk 4]
```diff
@@ -33,6 +37,13 @@
     @Override
     public void robotInit() {
         oi = new OI();
+
+        CameraServer.startAutomaticCapture();
+
+        CvSink cvSink = CameraServer.getVideo();
+
+        CvSource outputStream = CameraServer.putVideo("Back Camera", 720, 1280);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/Robot.java:45]: change to lowercase

File: src/main/java/frc/robot/commands/autonomous/GyroAutonomous.java [hunk 5]
```diff
@@ -24,12 +26,13 @@
     public GyroAutonomous(Drivetrain drivetrain) {
         super(
                 new ParallelCommandGroup(
+                        new InstantCommand(() -> IntakePlacer.getInstance().setServoAngle(90)).withTimeout(5),
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/autonomous/GyroAutonomous.java:29]: https://www.youtube.com/watch?v=oNXAsP1VbtU

File: src/main/java/frc/robot/commands/autonomous/GyroAutonomous.java [hunk 6]
```diff
@@ -24,12 +26,13 @@
...
                         new IntakeCargo(),
                         new SequentialCommandGroup(
                                 new MoveToCargo(drivetrain, MoveToCargo.CARGO_MOVE_VALUE),
                                 new DriveArcade(drivetrain, MoveToCargo.CARGO_MOVE_VALUE, () -> 0.0)
                         ).withInterrupt(IntakeToTransfer.getInstance()::getLimit)
-                ),
+                ).withTimeout(7),
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/autonomous/GyroAutonomous.java:35]: Just, I mean, come on...

File: src/main/java/frc/robot/commands/autonomous/SimpleSix.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    public static final double DRIVE_UNTIL_HIT_HUB_TIMEOUT = 3;
+    public static final double RELEASE_CARGO_TIMEOUT = 3;
+
+    public SimpleSix(Drivetrain drivetrain) {
+        super(
+                new ParallelCommandGroup(
+                        new IntakeCargo(),
+                        new DriveArcade(drivetrain, MoveToCargo.CARGO_MOVE_VALUE, () -> 0.0)
+                                .withInterrupt(IntakeToTransfer.getInstance()::getLimit)
+                ).withTimeout(7),
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/autonomous/SimpleSix.java:26]: I'm not mad, I'm just disappointed

File: src/main/java/frc/robot/commands/autonomous/SimpleSix.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+                new DriveUntilHitHub(drivetrain).withTimeout(DRIVE_UNTIL_HIT_HUB_TIMEOUT),
+                new ReleaseCargo().withTimeout(RELEASE_CARGO_TIMEOUT),
+                new DriveArcade(Drivetrain.getInstance(), RETREAT_DRIVE_SPEED, RETREAT_DRIVE_ROTATE)
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/autonomous/SimpleSix.java:29]: change `Drivetrain.getInstance()` to `drivetrain`

File: src/main/java/frc/robot/subsystems/BustedDigitalInput.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
+package frc.robot.subsystems;
+
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.Timer;
+
+public class BustedDigitalInput extends DigitalInput {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/BustedDigitalInput.java:6]: First document what the hell this means and then I'll tell you why you're wrong

File: src/main/java/frc/robot/subsystems/IntakePlacer.java [hunk 10]
```diff
@@ -60,6 +63,7 @@ private IntakePlacer() {
      */
     @Override
     public boolean canMove(double speed) {
+        rootNamespace.putBoolean("Can Move", !(isDown() && speed < 0) && !(isUp() && speed > 0));
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/IntakePlacer.java:66]: lowercase

File: src/main/java/frc/robot/subsystems/IntakePlacer.java [hunk 11]
```diff
@@ -17,9 +19,10 @@
 
     public static final double MAX_SPEED = 0.5;
     public static final double MIN_SPEED = -0.1;
-    public static final double SERVO_START_ANGLE = 0;
-    public static final double SERVO_TARGET_ANGLE = 20;
+    public static final double SERVO_START_ANGLE = 110;
+    public static final double SERVO_TARGET_ANGLE = 0;
 
+    public final Supplier<Double> TARGET = rootNamespace.addConstantDouble("TARGET", 90);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/IntakePlacer.java:25]: what's the difference between `SERVO_TARGET_ANGLE` and `TARGET`? Give better names.

File: src/main/java/frc/robot/commands/MoveToCargo.java [hunk 12]
```diff
@@ -9,7 +9,7 @@
 
 public class MoveToCargo extends DriveArcadeWithPID {
```
Comment (yutytuty) [inline src/main/java/frc/robot/commands/MoveToCargo.java:10]: add docs

File: src/main/java/frc/robot/commands/autonomous/GyroAutonomous.java (hunk 6 above)
Comment (yutytuty) [reply src/main/java/frc/robot/commands/autonomous/GyroAutonomous.java:35]: What he means is, make the time smaller

File: src/main/java/frc/robot/commands/autonomous/SimpleSix.java (hunk 7 above)
Comment (yutytuty) [reply src/main/java/frc/robot/commands/autonomous/SimpleSix.java:26]: ditto

File: src/main/java/frc/robot/subsystems/BustedDigitalInput.java (hunk 9 above)
Comment (yutytuty) [inline src/main/java/frc/robot/subsystems/BustedDigitalInput.java:6]: Is this class needed after mechanical fixed the limit?

File: src/main/java/frc/robot/subsystems/ClimberWinch.java [hunk 13]
```diff
@@ -38,6 +38,8 @@ public static ClimberWinch getInstance() {
 
     private ClimberWinch(CANSparkMax leftWinch, CANSparkMax rightWinch) {
         super(DOWN_SPEED, UP_SPEED, "climber winch", leftWinch, rightWinch);
+        leftWinch.setIdleMode(CANSparkMax.IdleMode.kBrake);
+        rightWinch.setIdleMode(CANSparkMax.IdleMode.kBrake);
```
Comment (yutytuty) [inline src/main/java/frc/robot/subsystems/ClimberWinch.java:42]: remove these two lines. They do not help.

**Review by yutytuty — DISMISSED** [review]

File: src/main/java/frc/robot/commands/autonomous/GyroAutonomous.java (hunk 6 above)
Comment (Yoel-Perman-Brilliant) [reply src/main/java/frc/robot/commands/autonomous/GyroAutonomous.java:35]: I meant magic but yea that too

File: src/main/java/frc/robot/OI.java [hunk 14]
```diff
@@ -39,10 +39,39 @@ public OI() {
...
-        xbox.getGreenButton().whenPressed(new MoveGenericSubsystem(climberWinch, ClimberWinch.DOWN_SPEED));
-        xbox.getYellowButton().whenPressed(new MoveGenericSubsystem(climberWinch, ClimberWinch.UP_SPEED));
+        //intakes two cargos
+        xbox.getLBButton().whenPressed(new ParallelCommandGroup(
+                new MoveGenericSubsystem(roller, IntakeRoller.MIN_SPEED),
+                new MoveGenericSubsystem(intakeToTransfer, IntakeToTransfer.SPEED),
+                new MoveGenericSubsystem(transfer, transfer.MOVE_SPEED).withInterrupt(transfer::getEntranceSensor)
+        ).withInterrupt(() -> (intakeToTransfer.getLimit()) && transfer.getEntranceSensor()));
+
+        xbox.getButtonStart().whenPressed(new MoveGenericSubsystem(IntakePlacer.getInstance(), 0) {
```
Comment (omri00) [inline src/main/java/frc/robot/OI.java:49]: Either use getInstance everywhere or a local var everywhere

File: src/main/java/frc/robot/OI.java [hunk 15]
```diff
@@ -39,10 +39,39 @@ public OI() {
...
+            @Override
+            public boolean isFinished() {
+                return true;
+            }
+        });
+
+        xbox.getGreenButton().whileHeld(new MoveGenericSubsystem(climberWinch, ClimberWinch.DOWN_SPEED) {
+            @Override
```
Comment (tal-sitton) [inline src/main/java/frc/robot/OI.java:57]: remove

File: src/main/java/frc/robot/OI.java [hunk 16]
```diff
@@ -39,10 +39,39 @@ public OI() {
...
+            public boolean isFinished() {
+                return false;
+            }
+        });
+        xbox.getYellowButton().whileHeld(new MoveGenericSubsystem(climberWinch, ClimberWinch.UP_SPEED) {
+            @Override
```
Comment (tal-sitton) [inline src/main/java/frc/robot/OI.java:63]: remove

File: src/main/java/frc/robot/OI.java (hunk 14 above)
Comment (tal-sitton) [inline src/main/java/frc/robot/OI.java:49]: add docs

File: src/main/java/frc/robot/OI.java [hunk 17]
```diff
@@ -39,10 +39,39 @@ public OI() {
...
+            public boolean isFinished() {
+                return false;
+            }
+        });
         xbox.getBlueButton().whenPressed(new MoveGenericSubsystem(climberWinch, 0));
```
Comment (tal-sitton) [inline src/main/java/frc/robot/OI.java:68]: add docs

File: src/main/java/frc/robot/Robot.java [hunk 18]
```diff
@@ -33,6 +37,13 @@
     @Override
     public void robotInit() {
         oi = new OI();
+
+        CameraServer.startAutomaticCapture();
+
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Robot.java:42]: why so many blank lines

File: src/main/java/frc/robot/commands/MoveToCargo.java [hunk 19]
```diff
@@ -7,24 +7,30 @@
 
 import java.util.function.Supplier;
 
+/**
+ * Moves to a cargo, while simultaneously aiming the robot to face the cargo.
+ */
 public class MoveToCargo extends DriveArcadeWithPID {
 
-    public static final Supplier<Double> CARGO_MOVE_VALUE = () -> 0.4;
+    public static final Supplier<Double> CARGO_MOVE_VALUE = () -> 0.53;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/commands/MoveToCargo.java:15]: change to `0.55`

**Review by omri00 — DISMISSED** [review]

File: src/main/java/frc/robot/subsystems/IntakePlacer.java [hunk 20]
```diff
@@ -60,7 +60,8 @@ private IntakePlacer() {
      */
     @Override
     public boolean canMove(double speed) {
-        return !(isDown() && speed < 0) && !(isUp() && speed > 0);
+        rootNamespace.putBoolean("can move", !(isDown() && speed < 0) && !(isUp() && speed > 0));
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/IntakePlacer.java:63]: remove

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/OI.java [hunk 21]
```diff
@@ -36,13 +37,36 @@ public OI() {
...
+        });
+
+        xbox.getGreenButton().whileHeld(new MoveGenericSubsystem(climberWinch, ClimberWinch.DOWN_SPEED));
+        xbox.getYellowButton().whileHeld(new MoveGenericSubsystem(climberWinch, ClimberWinch.UP_SPEED));
+
+        //stops the climberWinch
         xbox.getBlueButton().whenPressed(new MoveGenericSubsystem(climberWinch, 0));
-        xbox.getRedButton().whenPressed(new MoveGenericSubsystem(roller, 0));
+
+        xbox.getRedButton().whenPressed(new MoveGenericSubsystem(roller, 0) {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/OI.java:64]: add docs

File: src/main/java/frc/robot/subsystems/IntakePlacer.java [hunk 22]
```diff
@@ -60,6 +60,7 @@ private IntakePlacer() {
      */
     @Override
     public boolean canMove(double speed) {
+        rootNamespace.putBoolean("can move", !(isDown() && speed < 0) && !(isUp() && speed > 0));
```
Comment (tal-sitton) [inline src/main/java/frc/robot/subsystems/IntakePlacer.java:63]: remove

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/OI.java [hunk 23]
```diff
@@ -36,13 +37,37 @@ public OI() {
...
+
+        //stops the intakePlacer
+        xbox.getButtonStart().whenPressed(new MoveGenericSubsystem(intakePlacer, 0) {
+            @Override
+            public boolean isFinished() {
+                return true;
+            }
+        });
+
+        xbox.getGreenButton().whileHeld(new MoveGenericSubsystem(climberWinch, ClimberWinch.DOWN_SPEED));
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/OI.java:58]: change to whenPressed

File: src/main/java/frc/robot/OI.java [hunk 24]
```diff
@@ -36,13 +37,37 @@ public OI() {
...
+        xbox.getYellowButton().whileHeld(new MoveGenericSubsystem(climberWinch, ClimberWinch.UP_SPEED));
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/OI.java:59]: ditto

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

**Review by OfriRosenbaum — APPROVED** [review]

**Review by tal-sitton — APPROVED** [review]

**Pull request merged**